sdk/tests/core/test_az_mqtt5_connection.c

```diff
@@ -434,28 +668,917 @@ static void test_az_mqtt5_connection_open_idle_success(void** state)
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_reconnecttimeout(&mock_mqtt, AZ_MQTT5_CONNACK_UNSPECIFIED_ERROR);
+
+  assert_int_equal(az_mqtt5_connection_close(&test_connection), AZ_OK);
+
+  assert_int_equal(ref_conn_rsp_ok, 0);
+  assert_int_equal(ref_conn_req, 1);
+  assert_int_equal(ref_disconn_req, 1);
+}
+
+// Testing that when reconnecting, receiving a disconnect response from the broker leads to the
+// connection going into an idle state.
+static void test_az_mqtt5_connection_reconnectimeout_disconnect_rsp_success(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_reconnecttimeout(&mock_mqtt, AZ_MQTT5_CONNACK_UNSPECIFIED_ERROR);
+
+  assert_int_equal(
+      az_mqtt5_inbound_disconnect(
+          &mock_mqtt,
+          &(az_mqtt5_disconnect_data){ .disconnect_requested = true,
+                                       .tls_authentication_error = false }),
+      AZ_OK);
+
+  assert_int_equal(ref_disconn_rsp, 2);
+  assert_int_equal(ref_conn_rsp_ok, 0);
+  assert_int_equal(ref_disconn_req, 0);
+  assert_int_equal(ref_conn_req, 1);
+}
+
+// Testing that on an unsuccessful connection attempt due to an invalid certificate, the
+// credentials are swapped.
+static void test_az_mqtt5_connection_reconnecttimeout_credential_swap_success(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_reconnecttimeout(&mock_mqtt, AZ_MQTT5_CONNACK_NOT_AUTHORIZED);
+  _az_mqtt5_connection_reconnecttimeout_connecting(&test_connection);
+
+  assert_int_equal(test_connection._internal.client_certificate_index, 1);
+  assert_int_equal(ref_timeout, 1);
+}
+
+// Testing that when reconnecting, if the retry count is exhausted, the connection goes into an
+// idle state.
```

Comment:
From the puml, doesn't retry count exhausted go to faulted?

Reply:
No, I realize that the puml has not updated, if you look at the .puml file it shows the updated retry count to idle transition but not the rendered image